File: easycloud-loader/src/main/java/dev/easycloud/service/EasyCloudLoader.java
package dev.easycloud.service;

import dev.easycloud.service.update.LoaderUpdateProvider;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.Arrays;

@SuppressWarnings("ResultOfMethodCallIgnored")
@Slf4j
public final class EasyCloudLoader {

    @SneakyThrows
    public static void main(String[] args) {
        var local = Path.of("local");
        var resources = Path.of("resources");
        var libraries = resources.resolve("libraries");
        local.toFile().mkdirs();
        resources.toFile().mkdirs();
        libraries.toFile().mkdirs();

        copyFile("EasyCloudUpdater.class", libraries.resolve("EasyCloudUpdater.class"));

        if(Arrays.stream(args).toList().stream().anyMatch(arg -> arg.equals("-Dauto.updates=true"))) {
            new LoaderUpdateProvider();
        }
        new DependencyLoader();

        copyFile("easycloud-plugin.jar", resources.resolve("easycloud-plugin.jar"));
        copyFile("easycloud-api.jar", libraries.resolve("dev.easycloud.service-impl-stable.jar"));

        copyFile("easycloud-cluster.jar", Path.of("easycloud-cluster.jar"));

        var thread = processThread();

        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            if(thread.isAlive()) {
                thread.interrupt();
            }
        }));

        while (true) {
            if(!thread.isAlive()) {
                System.exit(0);
                break;
            }
        }
    }

    private static @NotNull Thread processThread() {
        var thread = new Thread(() -> {
            try {
                var fileArg = "easycloud-cluster.jar;resources/libraries/*;";
                if(!isWindows()) {
                    fileArg = fileArg.replace(";", ":");
                }

                var process = new ProcessBuilder("java", "-Xms512M", "-Xmx512M", "--enable-native-access=ALL-UNNAMED", "-cp", fileArg, "dev.easycloud.service.EasyCloudBootstrap")
                        .redirectOutput(ProcessBuilder.Redirect.INHERIT)
                        .redirectError(ProcessBuilder.Redirect.INHERIT)
                        .redirectInput(ProcessBuilder.Redirect.INHERIT)
                        .start();

                process.waitFor();
            } catch (IOException | InterruptedException exception) {
                if(exception instanceof InterruptedException) {
                    Thread.currentThread().interrupt();
                    System.exit(1);
                    return;
                }

                throw new RuntimeException(exception);
            }
        });
        thread.setDaemon(false);
        thread.start();
        return thread;
    }

    private static boolean isWindows() {
        return System.getProperty("os.name").toLowerCase().contains("win");
    }

    private static void copyFile(String name, Path path) {
        try {
            var file = ClassLoader.getSystemClassLoader().getResourceAsStream(name);
            if (file == null) {
                throw new RuntimeException("Resource " + name + " not found!");
            }
            Files.copy(file, path, StandardCopyOption.REPLACE_EXISTING);
        } catch (IOException exception) {
            throw new RuntimeException(exception);
        }
    }
}

File: easycloud-loader/src/main/java/dev/easycloud/service/EasyCloudLoader.kt
package dev.easycloud.service

import dev.easycloud.service.dependency.DependencyLoader
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardCopyOption
import java.util.*
import kotlin.io.path.createDirectory
import kotlin.io.path.deleteIfExists
import kotlin.io.path.exists
import kotlin.system.exitProcess

class EasyCloudLoader {

    fun load() {
        val localPath = Paths.get("local")
        val resourcesPath = Paths.get("resources")
        val librariesPath = resourcesPath.resolve("libs")
        val modulesPath = Paths.get("modules")

        // Create directories if they do not exist
        localPath.takeIf { !it.exists() }?.createDirectory()
        resourcesPath.takeIf { !it.exists() }?.createDirectory()
        librariesPath.takeIf { !it.exists() }?.createDirectory()
        modulesPath.takeIf { !it.exists() }?.createDirectory()

        // Check if the loader-patcher.jar file exists
        if (Files.exists(Paths.get("loader-patcher.jar"))) {
            println("Patching could have been failed.")
            Files.delete(Paths.get("loader-patcher.jar"))
        }

        // Load the EasyCloudCluster
        mapOf(
            Pair("easycloud-patcher", librariesPath.resolve("dev.easycloud.patcher")),
            Pair("easycloud-service", resourcesPath.resolve("easycloud-service")),
            Pair("easycloud-api", librariesPath.resolve("dev.easycloud.api")),
            Pair("easycloud-cluster", Paths.get("easycloud-cluster")),
            Pair("bridge-module", modulesPath.resolve("bridge-module")),
        ).forEach {
            this.copyFile("${it.key}.jar", Paths.get("${it.value}.jar"))
        }

        DependencyLoader().load(librariesPath)

        val thread = Thread {
            try {
                var fileArg = "easycloud-cluster.jar;resources/libs/*;"
                if (!System.getProperty("os.name").lowercase(Locale.getDefault()).contains("win")) {
                    fileArg = fileArg.replace(";", ":")
                }
                val process = ProcessBuilder(
                    "java",
                    "-Xms512M",
                    "-Xmx512M",
                    "--enable-native-access=ALL-UNNAMED",
                    "-cp",
                    fileArg,
                    "dev.easycloud.service.EasyCloudBoot"
                ).redirectOutput(ProcessBuilder.Redirect.INHERIT)
                    .redirectError(ProcessBuilder.Redirect.INHERIT)
                    .redirectInput(ProcessBuilder.Redirect.INHERIT)
                    .start()

                process.waitFor()
            } catch (exception: IOException) {
                throw RuntimeException(exception)
            } catch (_: InterruptedException) {
                Thread.currentThread().interrupt()
                exitProcess(1)
            }
        }
        thread.setDaemon(false)
        thread.start()

        Runtime.getRuntime().addShutdownHook(Thread {
            if (thread.isAlive) thread.interrupt()
        })

        while (true) {
            if (!thread.isAlive) exitProcess(0)
        }
    }

    private fun copyFile(name: String, destination: Path) {
        val file = ClassLoader.getSystemClassLoader().getResourceAsStream(name)
            ?: throw IllegalArgumentException("Resource $name not found")
        Files.copy(file, destination, StandardCopyOption.REPLACE_EXISTING)
    }
}